File: Data Structure/Lab 9/ReverseList.c
//WAP to reverse a linked list.

#include<stdio.h>
#include<stdlib.h>

struct Node
{
    int info;
    struct Node *link;
};

struct Node *first;

void insertAtFirst(int x){
    struct Node *newNode = (struct Node*)malloc(sizeof(struct Node*));

    if(newNode == NULL){
        printf("Memory allocation failed!List is Empty!!");
        return;
    }

    newNode->info = x;
    newNode->link = first;

    first = newNode;
}

void display(){
    struct Node *save = first;

    if(first == NULL){
        printf("Empty List!!");
        return;
    }

    while(save != NULL){
        printf("%d->",save->info);
        save = save->link;
    }
    printf("NULL");
}

void reverseList(){
    struct Node *prev = NULL;
    struct Node *save = first;
    struct Node *next = NULL;

    while(save != NULL){
        next = save->link;
        save->link = prev;

        prev = save;
        save = next;

    }
    first = prev;
}

void main(){

    insertAtFirst(9);
    insertAtFirst(8);
    insertAtFirst(7);
    insertAtFirst(6);

    display();

    reverseList();

    display();
}

File: Data Structure(Java)/Lab 7/ReverseList.java
//WAP to reverse a linked list.

class Node{
    int info;
    Node link;

    Node(int info){
        this.info = info;
        this.link = null;
    }
}

public class ReverseList {
    
static Node first;

static void insertAtFirst(int x){
    Node newNode = new Node(x);

    newNode.info = x;
    newNode.link = first;

    first = newNode;
}

static void displayNodes(){
    if(first == null){
        System.err.println("Empty List!!");
        return;
    }

    Node save = first;

    while(save != null){
        System.out.print(save.info+"->");
        save = save.link;
    }
    System.out.println("null");
}

static void reverseList(){
    Node prev = null;
    Node save = first;
    Node next = null;

    while(save != null){
        next = save.link;
        save.link = prev;

        prev = save;
        save = next; 
    }
    first = prev;
}

    public static void main(String[] args) {
        insertAtFirst(9);
        insertAtFirst(8);
        insertAtFirst(7);
        insertAtFirst(6);

        System.out.println("Original List:");
        displayNodes();

        reverseList();

        System.out.println("Reversed List:");
        displayNodes();
    }
}
